### DISCUSSION TITLE
Caching options for stateless environments

### DISCUSSION THREAD

**Discussion by davidwallacejackson** [dagger/dagger#3359] [General]
This is a follow-up on a conversation I started in today's Community Call about caching. For those who weren't on the call, my general question is, "what's Dagger 0.3's caching story for CI?" There's a lot that Dagger could do to make caching better/easier for all users, but I'd at least like to see the ability to do `--cache-from` equivalent to a standard Docker build.

I'll lay out an example here of how I use `--cache-from` in my own CI builds at present. I think this makes the power-user use case clear, but it also suggests a path to a generalizable cache experience that works for most users most of the time -- and without requiring expensive or tricky cloud resources beyond a standard container registry!

Using git to select cache sources from a container registry
----------------------------------------------------------

Every Docker build in my CI pipeline runs through a little Python wrapper. Here's a very condensed version of that wrapper...

```python
import subprocess

def exec_read(cmd: str) -> str:
    try:
        proc = subprocess.run(cmd, shell=True, check=True, capture_output=True)
    except subprocess.CalledProcessError as e:
        print(e.stdout)
        print(e.stderr)
        raise e

    return proc.stdout.decode("utf-8").rstrip()

DIRTY = "-dirty" if exec_read("git status -s") != "" else ""

GIT_SHA = exec_read("git describe --match= --always --abbrev=40")
GIT_SHA_DIRTY = GIT_SHA + DIRTY  # will be $GIT_SHA if repository is not dirty
GIT_BRANCH = exec_read("git rev-parse --abbrev-ref HEAD").replace("/", "-")
GIT_BRANCH_DIRTY = GIT_BRANCH + DIRTY  # will be $GIT_BRANCH if repository is not dirty
GIT_PARENT_SHA = exec_read("git describe --match= --always --abbrev=40 HEAD^")

WRITE_TAGS = [GIT_SHA_DIRTY, GIT_BRANCH_DIRTY]

CACHE_FROM_TAGS = [
    GIT_SHA_DIRTY,
    GIT_SHA,
    GIT_PARENT_SHA,
    GIT_BRANCH,
    "master",
]

def build(
    image: str,
    context_path: str,
    dockerfile: str,
) -> str:
    command = f"docker buildx build {context_path}"
    command += f" --file={dockerfile} --cache-to=type=inline"
    command += " \ \n  --push"

    for tag in CACHE_FROM_TAGS:
          command += f" \ \n  --cache-from={image}:{tag}"

    for tag in WRITE_TAGS:
        command += f" \ \n  --tag={image}:{tag}"


    return command


print(build("my-image", ".", "Dockerfile"))
```

You can try running this in any git repository -- it'll get you something like (running from branch `new-branch`)

```
docker buildx build . --file=Dockerfile --cache-to=type=inline \ 
  --push \ 
  --cache-from=my-image:e41621c82becdcfb48559f6f499cedb522e5eebd \ 
  --cache-from=my-image:e41621c82becdcfb48559f6f499cedb522e5eebd \ 
  --cache-from=my-image:0cbb072eff8db991f95928543e442b5dd43a697d \ 
  --cache-from=my-image:new-branch \ 
  --cache-from=my-image:master \ 
  --tag=my-image:e41621c82becdcfb48559f6f499cedb522e5eebd \ 
  --tag=my-image:new-branch
```

Or, if you'd changed something...

```
docker buildx build . --file=Dockerfile --cache-to=type=inline \ 
  --push \ 
  --cache-from=my-image:e41621c82becdcfb48559f6f499cedb522e5eebd-dirty \ 
  --cache-from=my-image:e41621c82becdcfb48559f6f499cedb522e5eebd \ 
  --cache-from=my-image:0cbb072eff8db991f95928543e442b5dd43a697d \ 
  --cache-from=my-image:new-branch \ 
  --cache-from=my-image:master \ 
  --tag=my-image:e41621c82becdcfb48559f6f499cedb522e5eebd-dirty \ 
  --tag=my-image:new-branch-dirty
```

Obviously, there's a lot of other stuff the full wrapper has to handle -- updating unpinned dependencies was a big one, and super interesting, but this post is already novel-length so I'll elide that for now. But in broad terms, I think this is a sensible caching strategy for almost any project, and I could imagine a future version of Dagger doing something like this by default.

Of course, there's still the cost of downloading and extracting the cache. But...

...you often don't need to download or extract the cache (if you write sensible Dockerfiles)
------------------------------------------------------------------------------------------

Here's a hypothetical Dockerfile

```Dockerfile
# syntax=docker/dockerfile:1.4

FROM ubuntu

# imagine some heavy dependencies here
RUN apt-get update && apt-get install curl

# In this example the binary is built outside of docker, because that's
# what we currently do -- but obviously it would be better to build it
# inside of buildkit and take advantage of the caching here, and that's
# part of what I want to accomplish with Dagger
COPY --link ./app_binary /app_binary

ENTRYPOINT ["/app_binary"]
```

Thanks to `COPY --link`, on a branch that changes the binary but not the base image or the apt dependencies, buildkit can solve this whole thing without actually needing to pull the contents of the cache.

Obviously, this is never going to be strictly _better_ than having a locally-available high-speed filesystem cache, and if we moved more of the build process into buildkit we'd have to pull layers here so that we had the dependencies to build the binary. But this strategy already gets you performance that's at least on par with what you can get out of any popular CI system. Combined with Dagger's ability to skip build steps that haven't changed, it'd be a particularly big win for monorepos.

**Comment by jpadams** [comment]
@davidwallacejackson Thanks for the writeup and sharing your wrapper script! Also, I just learned a [new word](https://www.merriam-webster.com/dictionary/elide)! 🙏 

I'll get an update and link relevant issues on `cache-to`, `cache-from` type support.

**Reply by mircubed** [reply]
@davidwallacejackson I created this issue, so we can start to track. I mentioned your hackathon timeline as well to see if we can help move this forward faster -https://github.com/dagger/dagger/issues/3422
